### ISSUE TITLE
Propagation of strides for tasklet indirection (was working on v0.14.4)

### ISSUE THREAD

**Issue by edopao** [spcl/dace#1486]
This issue was initially reported as a PR see #1479.

This issue arises in gt4py DaCe backend when trying to uplift dace from v0.14.4 to v0.15.1. Note that this issue affects gt4py cartesian, which is the legacy version under maintenance. This issue is not related to the ongoing development for ICON project.

The issue is about tasklets which contain array access. More specifically, when an array slice is passed as argument to the tasklet. I was able to reproduce this issue in a test case, which is available on [spcl/edopao:bug-cpp-strides](https://github.com/edopao/dace/tree/bug-cpp-strides)

The arrays in the nested-SDFG are defined as:
```
sdfg.add_array(
    'vec_field', (W, N), dc.float64, strides=(vec_stride_W, vec_stride_d0)
)
sdfg.add_array(
    'mat_field', (1, N, N), dc.float64, strides=(mat_stride_W, mat_stride_d0, mat_stride_d1)
)
```

The original tasklet in gt4py cartesian looked like:
```
for i in range({N}):
    _mat[i, i] = _vec[i] + dace.float64(1)
```

This was working with dace v0.14.4, it fails on v0.15.1 with exception on number of strides for `_mat` access:
```
if isinstance(visited_slice, ast.Tuple):
    # If slice is multi-dimensional and writes to array with more than 1 elements, then:
    # - Assume this is indirection (?)
    # - Soft-squeeze the slice (remove unit-modes) to match the treatment of the strides above.
    if target not in self.constants:
desc = self.sdfg.arrays[dname]
if isinstance(desc, data.Array) and data._prod(desc.shape) != 1:
    elts = [e for i, e in enumerate(visited_slice.elts) if desc.shape[i] != 1]
    else:
elts = visited_slice.elts
    if len(strides) != len(elts):
>       raise SyntaxError('Invalid number of dimensions in expression (expected %d, '
  'got %d)' % (len(strides), len(elts)))
E       SyntaxError: Invalid number of dimensions in expression (expected 2, got 1)

dace/codegen/targets/cpp.py:1120: SyntaxError
```

I already got some suggestion from @tbennun about using all explicit indices for array access in tasklet, so I changed it into:
```
for i in range({N}):
    _mat[0, i, i] = _vec[0, i] + dace.float64(1)
```

but I get this exception, now on `_vec` access:
```
    if len(strides) != len(elts):
>       raise SyntaxError('Invalid number of dimensions in expression (expected %d, '
  'got %d)' % (len(strides), len(elts)))
E       SyntaxError: Invalid number of dimensions in expression (expected 1, got 2)
```

It will work if I change to this:
```
for i in range({N}):
    _mat[0, i, i] = _vec[i] + dace.float64(1)
```

Although it works, it is not clear why `_mat` should be accessed with all explicit indices, while `_vec` only requires index for the slice dimension. Either this is expected behavior, and I need to understand how to choose one case or the other. Or this is a bug in CPP code generation: [dace/codegen/targets/cpp.py](https://github.com/spcl/dace/blob/master/dace/codegen/targets/cpp.py)
```
if isinstance(visited_slice, ast.Tuple):
    # If slice is multi-dimensional and writes to array with more than 1 elements, then:
    # - Assume this is indirection (?)
    # - Soft-squeeze the slice (remove unit-modes) to match the treatment of the strides above.
    if target not in self.constants:
desc = self.sdfg.arrays[dname]
if isinstance(desc, data.Array) and data._prod(desc.shape) != 1:
    elts = [e for i, e in enumerate(visited_slice.elts) if desc.shape[i] != 1]
    else:
elts = visited_slice.elts
    if len(strides) != len(elts):
raise SyntaxError('Invalid number of dimensions in expression (expected %d, '
  'got %d)' % (len(strides), len(elts)))
```

I would kindly ask @alexnick83 to take a look since he was the last to modify this code (8981b81).

**Comment by alexnick83** [comment]
To confirm, the correct Python code is:
```python
for i in range({N}):
    _mat[0, i, i] = _vec[j, i] + dace.float64(1)
```
where `j` is an index in `[0, W)`.  
Is this correct? Should any of the other versions work correctly when executed in the CPython interpreter?

**Comment by alexnick83** [comment]
Maybe I misunderstood here. Does gt4py output this as Python code subsequently parsed by DaCe or does the code appear as-is inside a Tasklet?

**Comment by edopao** [comment]
The code appears inside a Tasklet, please check the test case on this branch: https://github.com/spcl/dace/compare/master...edopao:dace:bug-cpp-strides

**Comment by edopao** [comment]
This test case is reproducing an issue I see in a more complicated SDFG. The reason for accessing `_vec[0, i]` is that the input memlet to the tasklet looks like:
```
state.add_edge(
    state.add_access('vec_field'), None,
    tasklet, '_vec',
    dc.Memlet.simple('vec_field', f"_w, 0:{N}")
)
```

**Comment by alexnick83** [comment]
Before discussing the main issue, there is another bug. The generated SDFG is invalid because you pass from the outer to the inner SDFG `vec_field[i, 0:2]`. In that case, the data in the inner SDFG may have the shape `(1, 2)` or `(2,)` but definitely not `(W, 2)`. Validation should fail, but it doesn't.

The main issue is conflicting design, which needs to be clarified before proceeding with a solution. I am not sure how to fix this without going back-and-forth breaking legitimate indirection. I will write below my interpretation of the SDFG semantics, but this needs to be addressed in a DaCe meeting.

@edopao I suggest that you also arrange a separate meeting to discuss how to provide a proper solution for gt4py in a timely manner.

TL;DR: (personal opinion) We shouldn't have regular range accesses in Python-language Tasklets because this should be strictly reserved (for pragmatic reasons) for indirection. One can see that in the code in the code generator that creates the issue. Range accesses inside a tasklet are considered to be the result of indirection.

In SDFG semantics, Tasklets are supposed to express the most fine-grained computation possible, i.e., operations among **scalars**. AFAIK, nothing in the original semantic document allows for non-scalar accesses inside a Tasklet, meaning that all Memlets in the edges adjacent to a Tasklet should communicate scalar values. There is only one exception to the above, which is indirection, i.e., `x[y[i]`. I will not discuss the indirection case further here because it is a separate issue by itself. However, I have to note that indirection is the reason that Python Tasklets have their code "analyzed" during code generation in the first place. It is a pragmatic feature.

Of course, this analysis opens the way to further support generic array accesses in Python Tasklets. I find two problems with this approach:
- We don't have clear semantics for how this should work. To some degree, the implementation roughly follows connector semantics of nested SDFGs. However, nested SDFG connectors are backed by actual data descriptors. Tasklets have nothing, only names. This leads to the second problem.
- The supposed semantics are implied and interpreted directly in the code generator after discarding a lot of information.

My opinion is that we should have Python code with range accesses inside Python Tasklets. The code should be parsed normally. If there are issues with the generated representation by the frontend then the frontend should be improved instead. I would also like to note that "C++ Tasklets" are free to include whatever ranges accesses they want because their code is not parsed/analyzed. It is just copied/pasted to the output file.

If it is decided that general range accesses inside Python Tasklets should be supported, then we need (1) clear semantics, (2) the capability to make a distinction from indirection, and (3) a mechanism that allows clear interpretation at the SDFG level rather than the code generator.

**Comment by edopao** [comment]
I have updated the dace backend in gt4py cartesian according to the above recommendations: when tasklet performs array access, it should use all explicit indexes. Therefore, the memlet shape was modified to pass the full array shape.
Link to PR in gt4py repo for reference: https://github.com/GridTools/gt4py/pull/1410
